src/odbc/odbc.c

```diff
@@ -1453,52 +1506,93 @@ _SQLBindParameter(SQLHSTMT hstmt, SQLUSMALLINT ipar, SQLSMALLINT fParamType, SQL
+			drec->sql_desc_octet_length = cbValueMax;
+		drec->sql_desc_indicator_ptr = pcbValue;
+		drec->sql_desc_octet_length_ptr = pcbValue;
+		drec->sql_desc_data_ptr = (char *) rgbValue;
 
-	/* field IPD related fields */
-	ipd = stmt->ipd;
-	orig_ipd_size = ipd->header.sql_desc_count;
-	if (ipar > ipd->header.sql_desc_count && desc_alloc_records(ipd, ipar) != SQL_SUCCESS) {
-		desc_alloc_records(apd, orig_apd_size);
-		odbc_errs_add(&stmt->errs, "HY001", NULL);
-		ODBC_EXIT_(stmt);
-	}
-	drec = &ipd->records[ipar - 1];
+		if (fSqlType == SQL_SS_TABLE) {
+			/* For a table value, the pcbValue is a buffer containing the number of rows present */
+			/* Replace it with a pointer to the actual size of a TDS_TVP struct */
+			drec->sql_desc_octet_length_ptr = NULL;
 
-	drec->sql_desc_parameter_type = fParamType;
-	if (odbc_set_concise_sql_type(fSqlType, drec, 0) != SQL_SUCCESS) {
-		desc_alloc_records(ipd, orig_ipd_size);
-		desc_alloc_records(apd, orig_apd_size);
-		odbc_errs_add(&stmt->errs, "HY004", NULL);
-		ODBC_EXIT_(stmt);
-	}
-	if (is_numeric) {
-		drec->sql_desc_precision = cbColDef;
-		drec->sql_desc_scale = ibScale;
+			/* Use this the column type as a marker for us */
+			/* to free up memory allocated with odbc_alloc_table() */
+			drec->sql_desc_type = drec->sql_desc_concise_type = SQL_C_SS_TABLE;
+
+			/* Table types are strictly read-only */
+			if (fParamType != SQL_PARAM_INPUT) {
+				odbc_errs_add(&stmt->errs, "HY105", NULL);
+				ODBC_EXIT_(stmt);
+			}
+
+			tvp = odbc_alloc_table();
+			if ((drec->sql_desc_data_ptr = (char *) tvp) == NULL) {
+				odbc_errs_add(&stmt->errs, "HY001", NULL);
+				ODBC_EXIT_(stmt);
+			}
+
+			tvp->type_name = rgbValue;
+			tvp->type_name_len = cbValueMax == SQL_NTS ? strlen(rgbValue) : cbValueMax;
+		}
+
+		/* field IPD related fields */
+		ipd = stmt->ipd;
+		orig_ipd_size = ipd->header.sql_desc_count;
+		if (ipar > ipd->header.sql_desc_count && desc_alloc_records(ipd, ipar) != SQL_SUCCESS) {
+			desc_alloc_records(apd, orig_apd_size);
+			odbc_errs_add(&stmt->errs, "HY001", NULL);
+			ODBC_EXIT_(stmt);
+		}
+		drec = &ipd->records[ipar - 1];
+
+		drec->sql_desc_parameter_type = fParamType;
+		if (odbc_set_concise_sql_type(fSqlType, drec, 0) != SQL_SUCCESS) {
+			desc_alloc_records(ipd, orig_ipd_size);
+			desc_alloc_records(apd, orig_apd_size);
+			odbc_errs_add(&stmt->errs, "HY004", NULL);
+			ODBC_EXIT_(stmt);
+		}
+		if (is_numeric) {
+			drec->sql_desc_precision = cbColDef;
+			drec->sql_desc_scale = ibScale;
+		} else {
+			drec->sql_desc_length = cbColDef;
+		}
 	} else {
-		drec->sql_desc_length = cbColDef;
+		/* Columns of table types are strictly read-only */
+		if (fParamType != SQL_PARAM_INPUT) {
+			odbc_errs_add(&stmt->errs, "HY105", NULL);
+			ODBC_EXIT_(stmt);
+		}
+
+		tvp = (SQLTVP *) stmt->apd[stmt->param_focus - 1].records->sql_desc_data_ptr;
+		if (add_table_column(tvp, ipar, fParamType, fCType, fSqlType, cbColDef, ibScale, rgbValue, cbValueMax, pcbValue) != SQL_SUCCESS) {
```

Comment:
What happen if I keep binding the same column over and over? Not necessary to fix in this series.